cupy/statistics/histogram.py

```diff
@@ -212,7 +214,23 @@ def histogram(x, bins=10, range=None, weights=None, density=False):
 
     if weights is None:
         y = cupy.zeros(bin_edges.size - 1, dtype='l')
-        _histogram_kernel(x, bin_edges, bin_edges.size, y)
+        # CUB uses int for input size count
+        # TODO(leofang): support >= 2^31 bins?
+        if (cupy.cuda.cub_enabled
+                and x.size <= 0x7fffffff and bin_edges.size <= 0x7fffffff):
+            # Need to ensure the dtype of bin_edges as it's needed for
+            # the correction later
+            if isinstance(bins, cupy.ndarray):
+                bin_type = cupy.result_type(bin_edges, x)
+                if cupy.issubdtype(bin_type, cupy.integer):
+                    bin_type = cupy.result_type(bin_type, float)
+                bin_edges = bin_edges.astype(bin_type)
+            y = cub.device_histogram(x, bin_edges, y)
+            # CUB's upper bin boundary is exclusive for all bins, including
+            # the last bin, so we must correct it to comply with NumPy
+            y[-1] += cupy.count_nonzero(x == bin_edges[-1])
```

Comment:
At least for integer data you could instead increment bin_edges[-1] by 1 prior to calling `cub.device_histogram` above (except for corner case where adding 1 would cause overflow).

For floats, maybe cupy.nextafter could be used to increment the final bin?

Reply:
Thanks, @grlee77! I actually tried a few variants of your trick, but the gotcha for integers is incrementing the last bin would include extra, unwanted samples larger than the origin edge but smaller or equal to the new edge. Ex:
```python
x = [1, 2, 4, 5, 6, 7, 8]
bin_edge = [0, 3, 5, 7]  # <-- 8 would also be included if we were to increment the last bin edge by 1.
```
For floats, `nextafter` might do the job. Let me test it...

Reply:
@grlee77 I verify that your suggestion works! Take a loot at 5cdf34d.

Reply:
By this I meant the overhead due to `count_nonzero` is largely eliminated. It's more transparent when the input array is of moderate size.